File: lib/domain/field_cultivation/policies/field_cultivation_climate_crop_view_policy.rb
# frozen_string_literal: true

module Domain
  module FieldCultivation
    module Policies
      # 気象 UC: 計画に紐づく作物の閲覧可否（I/O なし）。公開計画は参照作物のみ。
      module FieldCultivationClimateCropViewPolicy
        module_function

        def view_allowed?(user:, crop_entity:, plan_type_public:)
          if plan_type_public
            crop_entity.is_reference == true
          else
            Domain::Shared::Policies::CropPolicy.view_allowed?(
              user,
              is_reference: crop_entity.is_reference,
              user_id: crop_entity.user_id
            )
          end
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field_cultivation/policies/field_cultivation_climate_crop_view_policy.rs
use crate::field_cultivation::dtos::ClimateCropEntity;
use crate::shared::policies::crop_policy;
use crate::shared::user::User;

pub fn view_allowed(
    user: Option<&User>,
    crop_entity: &ClimateCropEntity,
    plan_type_public: bool,
) -> bool {
    if plan_type_public {
        crop_entity.is_reference
    } else {
        let Some(user) = user else {
            return false;
        };
        crop_policy::view_allowed(user, crop_entity.is_reference, crop_entity.user_id)
    }
}

#[cfg(test)]
mod policies_field_cultivation_climate_crop_view_policy_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/field_cultivation/policies_field_cultivation_climate_crop_view_policy_test.rs"));
}
